07_Agents/weather_service_agent/00_Function_DIY_Agents.ipynb

```diff
@@ -0,0 +1,1294 @@
+   "outputs": [],
+   "source": [
+    "from langchain import PromptTemplate\n",
+    "\n",
+    "TOOL_TEMPLATE = \"\"\"\\\n",
+    "Your job is to formulate a solution to a given <user-request> based on the instructions and tools below.\n",
+    "\n",
+    "Use these Instructions: \n",
+    "1. In this environment you have access to a set of tools and functions you can use to answer the question.\n",
+    "2. You can call the functions by using the <function_calls> format below.\n",
+    "3. Only invoke one function at a time and wait for the results before invoking another function.\n",
+    "4. The Results of the function will be in xml tag <function_results>. Never make these up. The values will be provided for you.\n",
+    "5. Only use the information in the <function_results> to answer the question.\n",
+    "6. Once you truly know the answer to the question, place the answer in <answer></answer> tags. Make sure to answer in a full sentence which is friendly.\n",
+    "7. Never ask any questions\n",
+    "\n",
+    "<function_calls>\n",
+    "<invoke>\n",
+    "<tool_name>$TOOL_NAME</tool_name>\n",
+    "<parameters>\n",
+    "<$PARAMETER_NAME>$PARAMETER_VALUE</$PARAMETER_NAME>\n",
+    "...\n",
+    "</parameters>\n",
+    "</invoke>\n",
+    "</function_calls>\n",
+    "\n",
+    "Here are the tools available:\n",
+    "<tools>\n",
+    "{tools_string}\n",
+    "</tools>\n",
+    "\n",
+    "<user-request>\n",
+    "{user_input}\n",
+    "</user-request>\n",
+    "\n",
+    "Human: What is the first step in order to solve this problem?\n",
+    "\n",
+    "Assistant:\n",
+    "\"\"\"\n",
+    "TOOL_PROMPT = PromptTemplate.from_template(TOOL_TEMPLATE)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "98bdbf85",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "## Executing the RAG Workflow\n",
+    "\n",
+    "Armed with our prompt and structured tools, we can now write an orchestration function which will iteratively step through the logical tasks to answer a user question. In the cell below we use the `invoke_model` function to generate a response with Claude and the `single_agent_step` function to iteratively call tools when the LLM tells us we need to. The general flow works like this...\n",
+    "\n",
+    "1. The user enters an input to the application\n",
+    "2. The user input is merged with the original prompt and sent to the LLM to determine the next step\n",
+    "3. If the LLM knows the answer, it will answer and we are done. If not, go to next step 4.\n",
+    "4. The LLM will determine which tool to use to answer the question.\n",
+    "5. We will use the tool as directed by the LLM and retrieve the results.\n",
+    "6. We provide the results back into the original prompt as more context.\n",
+    "7. We ask the LLM the next step or if knows the answer.\n",
+    "8. Return to step 3.\n",
+    "\n",
+    "If this is a bit confusing do not panic, we will walk through this flow in an example shortly!"
```

Comment:
lets not use the word 'panic'.

Reply:
thanks and good catch, have adjusted the wordings